assets/backport/github/github.go

```diff
@@ -0,0 +1,336 @@
+	// Create a new branch off of the target branch.
+	err := c.createBranchFrom(ctx, baseBranchName, newBranchName)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	fmt.Printf("Created a new branch: %s.\n", newBranchName)
+
+	// Cherry pick commits.
+	err = c.cherryPickCommitsOnBranch(ctx, newBranchName, commits)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	fmt.Printf("Finished cherry-picking %v commits. \n", len(commits))
+	return newBranchName, nil
+}
+
+// CreatePullRequest creates a pull request.
+func (c *Client) CreatePullRequest(ctx context.Context, baseBranch string, headBranch string, titleAndBody string) error {
+	newPR := &go_github.NewPullRequest{
+		Title:               go_github.String(titleAndBody),
+		Head:                go_github.String(headBranch),
+		Base:                go_github.String(baseBranch),
+		Body:                go_github.String(titleAndBody),
+		MaintainerCanModify: go_github.Bool(true),
+	}
+	_, _, err := c.Client.PullRequests.Create(ctx, c.Organization, c.Repository, newPR)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// GetPullRequestMetadata gets the commit shas, title, and body for a pull request
+// associated with the passed in branch name.
+func (c *Client) GetPullRequestMetadata(ctx context.Context, number int) (commits []string, branchName string, err error) {
+	pull, _, err := c.Client.PullRequests.Get(ctx, c.Organization, c.Repository, number)
+	if err != nil {
+		return nil, "", trace.Wrap(err)
+	}
+	if pull.GetState() != backportPRState {
+		return nil, "", trace.Errorf("pull request %v is not closed", number)
+	}
+	if strings.TrimPrefix(pull.GetBase().GetRef(), branchRefPrefix) != masterBranchName {
+		return nil, "", trace.Errorf("pull request %v's base is not master", number)
+	}
+
+	commits, err = c.getPullRequestCommits(ctx, pull.GetNumber())
+	if err != nil {
+		return nil, "", trace.Wrap(err)
+	}
+	return commits, strings.TrimPrefix(pull.GetHead().GetRef(), branchRefPrefix), nil
+}
+
+// cherryPickCommitsOnBranch cherry picks a list of commits on a given branch.
+func (c *Client) cherryPickCommitsOnBranch(ctx context.Context, branchName string, commits []string) error {
+	branch, _, err := c.Client.Repositories.GetBranch(ctx, c.Organization, c.Repository, branchName, true)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	// Get the branch's HEAD.
+	headCommit, _, err := c.Client.Git.GetCommit(ctx,
+		c.Organization,
+		c.Repository,
+		branch.GetCommit().GetSHA())
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	for i := 0; i < len(commits); i++ {
+		cherryCommit, _, err := c.Client.Git.GetCommit(ctx, c.Organization, c.Repository, commits[i])
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		tree, sha, err := c.cherryPickCommit(ctx, branchName, cherryCommit, headCommit)
+		if err != nil {
+			defer func() {
+				refName := fmt.Sprintf("%s%s", branchRefPrefix, branchName)
+				c.Client.Git.DeleteRef(ctx, c.Organization, c.Repository, refName)
+			}()
```

Comment:
Can you explain what happens if you don't delete the ref upon error? Is this something the use could just cleanup?

If it is needed, don't ignore the error, if it's not nil, log it.

Reply:
The user can clean up/delete the branch if they wanted to. I just added this in the event there is an error the user can fix their pr (rebase or something) and re-run the tool without having to worry about manually deleting it themselves.

Reply:
It's not **needed** so I'll log it when there is an error and tell the user to manually delete the branch to run again. Deleting the created branch is necessary because if the user wants to run it again, the auto-created branch can't exist.